User facing docs (#118)
kyma-project/registry-cache: docs/user/README.md

> # Registry Cache Module User Documentation

Reviewer: ```suggestion
# Registry Cache Module
```

> This document describes how to configure the Registry Cache for your Kyma Runtime cluster.

Reviewer: ```suggestion
Learn how to configure the Registry Cache module for your Kyma runtime cluster.
```

## Table of Contents
- [Introduction](#introduction)
- [Providing Basic Configuration](#basic-config.md)
- [Providing Credentials for Upstream Repository](#upstream-credentials.md)
- [Advanced Configuration](#advanced-config.md)
- [Validation of Registry Cache Configuration](#validation.md)
- [Managing Registry Cache Configuration](#managing-registry-cache-configuration)

> - [Troubleshooting](#troubleshooting.md)

Reviewer: ```suggestion
```

Reviewer: Since we're preparing docs that will also be published on Help Portal, let's delete the toc. We won't have one in the HP version.

> ## Introduction

Reviewer: ```suggestion
## What Is Registry Cache?
```

> The Registry Cache Kyma module adds a caching layer for container image registries used in your BTP managed Kyma Runtimes.

Reviewer: ```suggestion
The Registry Cache Kyma module adds a caching layer for container image registries used in your SAP BTP, Kyma runtime instances.
```

## Providing Credentials for Upstream Repository

If the upstream registry requires authentication, create a Kubernetes Secret in the same namespace as the `RegistryCacheConfig` resource and reference it in the **spec.secretReferenceName** field.
The Secret must be immutable and of type `generic`.

> ### Note:
> The credential Secret must exist on the cluster **before** applying the `RegistryCacheConfig` resource.

1. Set environment variables with the upstream registry credentials:

```bash
export USERNAME=<your username>
export PASSWORD=<your password>
```

2. Create the namespace if it doesn't exist:

```bash
kubectl create namespace test
```

3. Create an immutable Secret named `rc-secret` in the `test` namespace:

```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: rc-secret
  namespace: test
type: Opaque
immutable: true
data:
  username: $(echo -n $USERNAME | base64 | tr -d '\n')
  password: $(echo -n $PASSWORD | base64 | tr -d '\n')
EOF
```

> For [Google Artifact Registry](https://cloud.google.com/artifact-registry/docs/docker/authentication), use `_json_key` as the username and the service account key in JSON format as the password.

Reviewer 2: I would move the GAR part to a separate header/file. 

This paragraph should have a separate secret snippet

```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: rc-secret-gcr
  namespace: test
type: Opaque
immutable: true
data:
  username: _json_key
  password: <service account key in JSON format as the password>
EOF
```

# Registry Cache Module User Documentation

This document describes how to configure the Registry Cache for your Kyma Runtime cluster.

## Table of Contents
- [Introduction](#introduction)
- [Providing Basic Configuration](#basic-config.md)
- [Providing Credentials for Upstream Repository](#upstream-credentials.md)
- [Advanced Configuration](#advanced-config.md)
- [Validation of Registry Cache Configuration](#validation.md)
- [Managing Registry Cache Configuration](#managing-registry-cache-configuration)
- [Troubleshooting](#troubleshooting.md)

## Introduction
The Registry Cache Kyma module adds a caching layer for container image registries used in your BTP managed Kyma Runtimes.
This reduces outbound traffic to public registries, improving performance and reliability of image pulls.
It also supports access to private registries by allowing you to provide credentials for the caching layer to use when authenticating against those registries.

> > The Registry Cache feature is built on top of [Gardener's Registry Cache extension](https://gardener.cloud/docs/extensions/others/gardener-extension-registry-cache/registry-cache/configuration/).

Reviewer: ```suggestion

> ### Note:
> The Registry Cache module is built on top of [Gardener's Registry Cache extension](https://gardener.cloud/docs/extensions/others/gardener-extension-registry-cache/registry-cache/configuration/).
```

## Prerequisites

> - A managed Kyma Runtime instance running on the BTP platform.

Reviewer: ```suggestion
- A SAP BTP, Kyma runtime instance running on the BTP platform.
```

Reviewer: Is "running on the BTP platform" necessary when we name the instance properly?

> - Administrative access to the Kyma Runtime with kubeconfig and the `kubectl` tool.

Reviewer: ```suggestion
- Administrative access to the Kyma runtime with kubeconfig and the `kubectl` tool.
```

> - The Registry Cache module installed on your Kyma Runtime cluster.

Reviewer: ```suggestion
- The Registry Cache module installed on your Kyma cluster.
```

## Basic Configuration
`RegistryCacheConfig` is a namespace-scoped resource and can be created in any namespace.

> To configure the Registry Cache, create a `RegistryCacheConfig` custom resource. The following example uses the `test` namespace — create it first if it doesn't exist:

Reviewer: ```suggestion
To configure Registry Cache, create a `RegistryCacheConfig` custom resource (CR). The following example uses the `test` namespace — create it first if it doesn't exist:
```

```bash
kubectl create namespace test
```

```bash
kubectl create -f - <<EOF 
apiVersion: core.kyma-project.io/v1beta1
kind: RegistryCacheConfig
metadata:
  name: config1
  namespace: test
spec:
  upstream: docker.io
  volume:
    size: 100Gi
EOF
```

> Once applied, the Kyma Control Plane processes the resource and configures a caching layer for the specified upstream registry (in this case `docker.io`).

Reviewer: ```suggestion
Once applied, Kyma Control Plane (KCP) processes the resource and configures a caching layer for the specified upstream registry (in this case, `docker.io`).
```

> The `volume.size` field specifies the size of the persistent volume used to store cached images.

Reviewer: ```suggestion
The **volume.size** field specifies the size of the persistent volume used to store cached images.
```

You can create multiple `RegistryCacheConfig` resources to cache different upstream registries. Each resource must have a unique name, and each upstream registry must be unique across all resources in the cluster.

## Providing Credentials for Upstream Repository

> If the upstream registry requires authentication, create a Kubernetes Secret in the same namespace as the `RegistryCacheConfig` resource and reference it in the `spec.secretReferenceName` field.

Reviewer: ```suggestion
If the upstream registry requires authentication, create a Kubernetes Secret in the same namespace as the `RegistryCacheConfig` resource and reference it in the **spec.secretReferenceName** field.
```

> The secret must be immutable and of type `generic`.

Reviewer: ```suggestion
The Secret must be immutable and of type `generic`.
```

> **Note:**

Reviewer: ```suggestion
> ### Note:
```

> > The credential secret must exist on the cluster **before** applying the `RegistryCacheConfig` resource.

Reviewer: ```suggestion
> The credential Secret must exist on the cluster **before** applying the `RegistryCacheConfig` resource.
```

1. Set environment variables with the upstream registry credentials:

```bash
export USERNAME=<your username>
export PASSWORD=<your password>
```

2. Create the namespace if it doesn't exist:

```bash
kubectl create namespace test
```

> 3. Create an immutable secret named `rc-secret` in the `test` namespace:

Reviewer: ```suggestion
3. Create an immutable Secret named `rc-secret` in the `test` namespace:
```

```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: rc-secret
  namespace: test
type: Opaque
immutable: true
data:
  username: $(echo -n $USERNAME | base64 | tr -d '\n')
  password: $(echo -n $PASSWORD | base64 | tr -d '\n')
EOF
```

For [Google Artifact Registry](https://cloud.google.com/artifact-registry/docs/docker/authentication), use `_json_key` as the username and the service account key in JSON format as the password.
To base64-encode the service account key, run:

```bash
echo -nE $SERVICE_ACCOUNT_KEY_JSON | base64 | tr -d '\n'
```

> 4. Apply the Registry Cache configuration referencing the created secret:

Reviewer: ```suggestion
4. Apply the Registry Cache configuration referencing the created Secret:
```

```bash
kubectl create -f - <<EOF
apiVersion: core.kyma-project.io/v1beta1
kind: RegistryCacheConfig
metadata:
  name: config2
  namespace: test
spec:
  upstream: <protected registry URL>
  secretReferenceName: rc-secret
  volume:
    size: 100Gi
EOF
```

> **Note:**

Reviewer: ```suggestion
> ### Note:
```

> When using a private registry, the same credentials must be stored in **two** Kubernetes Secrets:

> > - The secret referenced in `spec.secretReferenceName` — used by the registry cache to authenticate against the upstream registry when pulling images to cache.

Reviewer: ```suggestion
> - The Secret referenced in **spec.secretReferenceName** — used by Registry Cache to authenticate against the upstream registry when pulling images to cache.
```

> > - An `imagePullSecret` on each workload — used by containerd to authenticate directly against the upstream registry as a fallback when the registry cache is unavailable.

Reviewer: ```suggestion
> - An `imagePullSecret` on each workload — used by containerd to authenticate directly against the upstream registry as a fallback when Registry Cache is unavailable.
```

>

> > Do not remove the `imagePullSecret` from your workloads when configuring credentials for the registry cache. If the cache is unavailable, containerd falls back to the upstream registry and requires the credentials directly.

Reviewer: ```suggestion
> Do not remove the `imagePullSecret` from your workloads when configuring credentials for Registry Cache. If the cache is unavailable, containerd falls back to the upstream registry and requires the credentials directly.
```

## Advanced Configuration

The following table describes all fields in the `RegistryCacheConfig` resource:

> | Field                          | Description                                                                                                                                                                                                                                                                                                                                                                                                            | Default value | Required |

Reviewer: In the table, I recommend the following changes:
- The Required column before the Description column
- Bold font for the field names (not code font)
- "None" instead of "N/A" (without quotation marks or any special formatting)

Author: Done

| `spec.garbageCollection.ttl`   | The time-to-live (TTL) for cached images. Images not accessed within this duration are eligible for garbage collection. Set to `0s` to disable garbage collection.                                                                                                                                                                                                                                                     | 168h (7 days) | No       |
| `spec.proxy.httpProxy`         | Proxy server for HTTP connections used by the registry cache.                                                                                                                                                                                                                                                                                                                                                          | N/A           | No       |

> | `spec.proxy.httpsProxy`        | Proxy server for HTTPS connections used by the registry cache.                                                                                                                                                                                                                                                                                                                                                         | N/A           | No       |

Reviewer: When referring to the module, pls use the following naming convention:
"Registry Cache" (capitalized, without "the") or "the Registry Cache module".

Author: Done

Author: Done

| `spec.http.tls`                | Indicates whether TLS is enabled for the HTTP server of the registry cache.                                                                                                                                                                                                                                                                                                                                            | true          | No       |

## Validation of Registry Cache Configuration

> After applying the `RegistryCacheConfig` resource, the registry cache webhook validates the configuration before it takes effect.

Reviewer: ```suggestion
After applying the `RegistryCacheConfig` resource, the Registry Cache webhook validates the configuration before it takes effect.
```

If the configuration is valid, the resource status transitions from `Pending` to `Ready` and the caching layer is configured.
If there are issues, the status transitions from `Pending` to `Error` and an error message is provided in the status conditions.

Example error message:
```
admission webhook "registrycacheconfig-v1beta1.kb.io" denied the request: spec.upstream: Invalid value: "dockerrrrr.io": upstream is not DNS resolvable
```

The following table describes the validation rules for each field:

> | Field                          | Validation                                                                                                                                         | Example |

Reviewer: Please see the comment above: https://github.com/kyma-project/registry-cache/pull/118/changes#r3458565681.

Reviewer: Please use the bold font for the field names.

### Listing Registry Cache Configurations

To list all `RegistryCacheConfig` resources across all namespaces, run:

```bash
kubectl get registrycacheconfig -A
```

To list resources in a specific namespace, run:

```bash
kubectl get registrycacheconfig -n <namespace>
```

### Deleting a Registry Cache Configuration

To delete a `RegistryCacheConfig` resource, run:

```bash
kubectl delete registrycacheconfig <name> -n <namespace>
```

For example:

```bash
kubectl delete registrycacheconfig config -n test
```

## Troubleshooting

The Registry Cache configuration is validated before being applied to the cluster. Invalid configuration will be rejected by the webhook.
If the configuration is valid but the Registry Cache setup fails on the KCP side, the `RegistryCacheConfig` resource status transitions to `Error` with an error message in the status conditions. In this case, contact the Kyma support team for assistance.

### Diagnosing Incorrect Credentials

The upstream registry does not return an explicit authentication error when credentials are wrong. Instead, image pulls fail with `404 manifest unknown`, which is indistinguishable from a missing image at the log level.

> If image pulls fail consistently with `404` errors and you know the image exists in the upstream registry, check the registry cache pod logs for the affected upstream:

Reviewer: ```suggestion
If image pulls fail consistently with `404` errors and you know the image exists in the upstream registry, check the Registry Cache Pod logs for the affected upstream:
```

```bash
kubectl logs -n kube-system -l app=registry-cache --tail=50
```

> To filter logs for a specific upstream, use the pod name pattern (pods are named after the upstream host):

Reviewer: ```suggestion
To filter logs for a specific upstream, use the Pod name pattern (Pods are named after the upstream host):
```

```bash
kubectl logs -n kube-system $(kubectl get pods -n kube-system -o name | grep registry-<upstream-host>) --tail=50
```

> A pull failure due to incorrect credentials looks like:

Reviewer: ```suggestion
A pull failure due to incorrect credentials looks similar to this one:
```
